speechbrain/lobes/models/discrete/SQ-Codec.py

```diff
@@ -0,0 +1,1497 @@
+        Sample rate for input audio (default is 16000).
+    dim_codebook : int, optional
+        Dimension of each codebook (default is 19683).
+    n_codebook : int, optional
+        Number of codebooks used (default is 4).
+    bw : float, optional
+        Bandwidth parameter (default is 2).
+    device : torch.device, optional
+        Device to run the model on (default is CPU).
+    clip_length : int, optional
+        Maximum clip length for processing (default is 450).
+
+    Example
+    -------
+    >>> model_hub = "Dongchao/UniAudio"
+    >>> save_path = "savedir"
+    >>> filename = "SQ-Codec.zip"
+    >>> config = "config.yaml"
+    >>> checkpoint = "ckpt_00190000.pth"
+    >>> model = SQCodec(model_hub, filename, save_path, config, checkpoint) # doctest: +SKIP
+    >>> audio = torch.randn(3, 16000)
+    >>> tokens, emb = model.encode(audio) # doctest: +SKIP
+    >>> tokens.shape # doctest: +SKIP
+    torch.Size([3, 200])
+    >>> emb.shape # doctest: +SKIP
+    torch.Size([3, 36, 50])
+    >>> rec = model.decode(tokens) # doctest: +SKIP
+    >>> rec.shape # doctest: +SKIP
+    torch.Size([3, 1, 16000])
+    """
+
+    def __init__(
+        self,
+        source,
+        filename,
+        save_path=None,
+        config="config.yaml",
+        checkpoint="ckpt_00190000.pth",
+        sample_rate=16000,
+        dim_codebook=19683,
+        n_codebook=4,
+        bw=2,
+        device=torch.device("cpu"),
+        clip_length=450,
+    ):
+        super(SQCodec, self).__init__()
+        self.config_path = os.path.join(
+            save_path, filename.split(".")[0], config
+        )
+        self.ckpt_path = os.path.join(
+            save_path, filename.split(".")[0], checkpoint
+        )
+        if not os.path.exists(self.config_path) and not os.path.exists(
+            self.ckpt_path
+        ):
+            download_and_extract(source, filename, save_path)
+        self.device = device
+        self.clip_length = clip_length
+
+        logging.info(
+            f"Using config {self.config_path} and model {self.ckpt_path}"
+        )
+        self.scalar_codec = self.build_codec_model(self.config_path).to(device)
+        self.sr = sample_rate
+        self.dim_codebook = dim_codebook
+        self.n_codebook = n_codebook
+        self.bw = bw
+        self.freq = self.n_codebook * 50
+        self.mask_id = self.dim_codebook * self.n_codebook
+
+    def build_codec_model(self, config):
+        """
+        Loads and builds the scalar codec model from the given configuration.
+
+        Parameters
+        ----------
+        config : str
+            Path to the configuration file.
+
+        Returns
+        -------
+        ScalarModel
+            The built scalar codec model loaded with weights from the checkpoint.
+        """
+        exp_model_config = OmegaConf.load(config)
+        scalar_codec = ScalarModel(**exp_model_config.generator.config)
+        parameter_dict = torch.load(self.ckpt_path)
```

Comment:
This operation looks like something that our Pretrainer class should be able to handle, it would be wiser to use it for clarity and our users habits.